### PR TITLE
add yearquarter, move date calculation from POSIXlt to C

### PR DESCRIPTION
**Pull request by ben-schwen**
Towards #649 and #4870

### CODE REVIEW

**Comment by ben-schwen** [comment]
## Benchmarks

### Setup
```r
set.seed(373L)
x = as.IDate(sample(seq(-25000, 45000), 1e7, TRUE))
```

### Comparison with existing functions
```r
bench::mark(
  data.table::yday(x),
  yday(x)
)
#   expression                  min   median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time
#   <bch:expr>             <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm>
# 1 data.table::yday(x)       1.09s    1.09s     0.914   495.9MB     1.83     1     2      1.09s
# 2 yday(x)                141.67ms 142.66ms     6.95     38.1MB     1.74     4     1   575.82ms

bench::mark(
  data.table::wday(x),
  wday(x)
)
#   expression                  min   median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time
#   <bch:expr>             <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm>
# 1 data.table::wday(x)     153.2ms  154.3ms      6.42    38.2MB     0        4     0      623ms
# 2 wday(x)                  58.8ms   59.4ms     16.9     38.1MB     2.11     8     1      475ms

bench::mark(
  data.table::mday(x),
  mday(x)
)
#   expression                  min   median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time
#   <bch:expr>             <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm>
# 1 data.table::mday(x)       1.11s    1.11s     0.901   457.8MB     1.80     1     2      1.11s
# 2 mday(x)                276.44ms  284.7ms     3.51     38.1MB     0        2     0    569.4ms

bench::mark(
  data.table::week(x),
  week(x)
)
#   expression                  min   median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time
#   <bch:expr>             <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm>
# 1 data.table::week(x)       1.15s    1.15s     0.870   495.9MB     1.74     1     2      1.15s
# 2 week(x)                168.16ms 168.49ms     5.94     38.1MB     0        3     0   505.36ms

bench::mark(
  data.table::month(x),
  month(x)
)
#   expression                   min   median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time
#   <bch:expr>              <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm>
# 1 data.table::month(x)       1.11s    1.11s     0.904   495.9MB     1.81     1     2      1.11s
# 2 month(x)                265.59ms 266.79ms     3.75     38.1MB     0        2     0   533.57ms

bench::mark(
  data.table::quarter(x),
  quarter(x)
)
#   expression                     min   median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time
#   <bch:expr>                <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm>
# 1 data.table::quarter(x)       1.13s    1.13s     0.884   495.9MB     1.77     1     2      1.13s
# 2 quarter(x)                218.03ms 218.37ms     4.56     38.1MB     0        3     0   658.21ms

bench::mark(
  data.table::year(x),
  year(x)
)
#   expression                  min   median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time
#   <bch:expr>             <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm>
# 1 data.table::year(x)       1.11s    1.11s     0.898   495.9MB     1.80     1     2      1.11s
# 2 year(x)                115.31ms 115.82ms     8.63     38.1MB     0        5     0   579.42ms
```

### Comparison with zoo
```r
bench::mark(
  zoo::as.yearmon(x),
  zoo::yearmon(yearmon(x))
)
#   expression                    min   median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time
#   <bch:expr>               <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm>
# 1 zoo::as.yearmon(x)          1.23s    1.23s     0.814     687MB     2.44     1     3      1.23s
# 2 zoo::yearmon(yearmon(x)) 348.14ms 362.54ms     2.76      153MB     1.38     2     1   725.08ms


bench::mark(
  zoo::as.yearqtr(x),
  zoo::yearqtr(yearqtr(x))
)
#   expression                    min   median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time
#   <bch:expr>               <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm>
# 1 zoo::as.yearqtr(x)          1.33s    1.33s     0.753     839MB     2.26     1     3      1.33s
# 2 zoo::yearqtr(yearqtr(x)) 294.21ms 296.24ms     3.38      153MB     0        2     0   592.48ms
```

File: R/IDateTime.R [hunk 1]
```diff
@@ -356,7 +356,13 @@ isoweek = function(x) {
...
-month   = function(x) as.POSIXlt(x)$mon + 1L
-quarter = function(x) as.POSIXlt(x)$mon %/% 3L + 1L
-year    = function(x) as.POSIXlt(x)$year + 1900L
+month   = function(x) convertDate(as.IDate(x), "month")
+quarter = function(x) convertDate(as.IDate(x), "quarter")
+year    = function(x) convertDate(as.IDate(x), "year")
+yearmon = function(x) convertDate(as.IDate(x), "yearmon")
+yearqtr = function(x) convertDate(as.IDate(x), "yearqtr")
 
+convertDate = function(x, type=c("yday", "wday", "mday", "week", "month", "quarter", "year", "yearmon", "yearqtr")) {
```
Comment (MichaelChirico) [inline R/IDateTime.R:365]: Q: should we export this? if so, `type` shouldn't get a default -- move the list of options into `match.arg()`.

File: inst/tests/tests.Rraw [hunk 2]
```diff
@@ -18702,3 +18704,14 @@ test(2234.7, DT[, min(.SD), by=c(.I)], data.table(I=1L:5L, V1=c(1L, 2L, 3L, 2L,
 test(2234.8, DT[, min(.SD), by=.I%%2L], error="by.*contains .I.*supported")  # would be nice to support in future; i.e. by odd/even rows, and by=(.I+1L)%/%2L for pairs of rows; i.e. any expression of .I
 test(2234.9, DT[, min(.SD), by=somefun(.I)], error="by.*contains .I.*supported")
 
+# move IDate calculating from POSIXlt to C, add yearquarter #649
+x = c("1111-11-11", "2019-01-01", "2019-02-28", "2019-03-01", "2019-12-31", "2020-02-29", "2020-03-01", "2020-12-31")
```
Comment (MichaelChirico) [inline inst/tests/tests.Rraw:18708]: test a date far in the future too?

File: src/idatetime.c [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+        double *ansp = REAL(ans);
+        for (int i=0; i < n; ++i) {
+            ansp[i] = convertDate_double(ix[i], ctype);
+        }
+    }
+    UNPROTECT(1);
+    return ans;
+}
+
+int convertDate_int(int x, datetype type)
```
Comment (MichaelChirico) [inline src/idatetime.c:51]: it looks like there's a lot of duplication between `convertDate_int` and `convertDate_double` -- could they be combined? C++ templates would help a lot here... `void convertSingleDate(int x, double *double_out, int *int_out, is_double, datetype type)` might work?

Otherwise, use more helper functions to unify shared logic across the two functions...

**Comment by MichaelChirico** [comment]
Improvement on memory efficiency looks great. Thanks!

File: src/idatetime.c (hunk 3 above)
Comment (ben-schwen) [reply src/idatetime.c:51]: Indeed. I also thought about either passing result pointers or using a struct where the results are stored, although the pointers seem favorable in terms of speed. Will rewrite it the next days.

File: inst/tests/tests.Rraw (hunk 2 above)
Comment (ben-schwen) [reply inst/tests/tests.Rraw:18708]: Shouldn't apply here since we always have days since 1970-01-01 or 2000-03-01 (in the calculation itself) but will add a testcase.

Comment (MichaelChirico) [reply inst/tests/tests.Rraw:18708]: agreed, mainly had it in mind as a regression test

File: R/IDateTime.R (hunk 1 above)
Comment (ben-schwen) [reply R/IDateTime.R:365]: No sure if we should export.

**Comment by ben-schwen** [comment]
@MichaelChirico I guess this PR also closes #2463?

**Comment by MichaelChirico** [comment]
> @MichaelChirico I guess this PR also closes #2463?

SGTM! Maybe add a benchmark against those too?

File: NEWS.md [hunk 4]
```diff
@@ -292,6 +292,8 @@
     # 2:     2    10
     ```
 
+40. Fast versions of `yearmon` and `yearqtr` are now available. Moreover, `yday`, `wday`, `mday`, `week`, `month` and `year` are now optimized regarding memory usage and speed by removing the `POSIXlt` dependency, [#649](https://github.com/Rdatatable/data.table/issues/649). Thanks to Matt Dowle for the request, and Benjamin Schwendinger for the PR.
```
Comment (MichaelChirico) [inline NEWS.md:295]: Maybe reword?

> New functions `yearmon()` and `yearqtr()` give .... These and also `yday`, ... are now optimized for memory and compute efficiency by removing the `POSIXlt` dependency...

**Review by MichaelChirico — APPROVED** [review]
Looking great, thanks!

**Comment by MichaelChirico** [comment]
looks like I can't "mark as resolved" since I commented in the wrong place 😵

**Comment by berg-michael** [comment]
I think this might cause a regression when passed missing values.

```
library(data.table)

yday(NA)
#> [1] 194
wday(NA)
#> [1] 3
mday(NA)
#> [1] 12
week(NA)
#> [1] 28
month(NA)
#> [1] 7
quarter(NA)
#> [1] 3
year(NA)
#> [1] 5881580
yearmon(NA)
#> [1] 5881581
yearqtr(NA)
#> [1] 5881581

data.table:::convertDate(NA_integer_, "yday")
#> [1] 194
data.table:::convertDate(NA_integer_, "year")
#> [1] 5881580
```

**Comment by jangorecki** [comment]
@berg-michael thank you. I filled new issue.
